### PR TITLE
Initial work on Manage Packages page (and Delete Account shared code)

### PR DESCRIPTION
**Pull request by chenriksson**
Updates PackageService.FindPackagesByOwner to include organization packages, which is used by both the ManagePackages and DeleteAccount views.

For Issue #4894:
- Displays organization packages
- Added logo next to Organization owners
- Remaining work: Filtering by owner will come in separate PR

For Issue #5016:
- Displays organization packages, and changes orphaned state to factor in organizations
- Added logo next to Organization owners
- Disables delete of organization and organization member accounts, until we have time to finish this

### CODE REVIEW

File: src/NuGetGallery/Extensions/UserExtensions.cs [hunk 1]
```diff
@@ -70,6 +71,21 @@ public static bool IsOwnerOrMemberOfOrganizationOwner(this User user, PackageReg
             return package.Owners.Any(o => user.KeyIsSelfOrOrganization(o.Key));
         }
 
+        /// <summary>
+        /// Get matching package owner keys for user, used when searching for packages.
+        /// </summary>
+        public static List<int> GetPackageOwnerKeys(this User user)
+        {
```
Comment (cristinamanum) [inline src/NuGetGallery/Extensions/UserExtensions.cs]: This method does not have anything to do with packages. Maybe a different name ?

File: src/NuGetGallery/Services/DeleteAccountService.cs [hunk 2]
```diff
@@ -66,15 +67,43 @@ AuthenticationService authService
...
                 {
                     Success = false,
-                    Description = string.Format(Strings.AccountDelete_AccountAlreadyDeleted, userToBeDeleted.Username),
+                    Description = string.Format(CultureInfo.CurrentCulture,
+                        Strings.AccountDelete_OrganizationDeleteNotImplemented,
+                        userToBeDeleted.Username),
                     AccountName = userToBeDeleted.Username
                 };
             }
+            else if (userToBeDeleted.Organizations.Any())
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/DeleteAccountService.cs]: This will need to be supported to be GDPR compliant. Could we have a task that tracks that we do not miss it?

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 3]
```diff
@@ -124,7 +124,6 @@ public virtual ActionResult DeleteRequest()
                 Packages = listPackageItems,
                 User = user,
                 AccountName = user.Username,
-                HasOrphanPackages = listPackageItems.Any(p => p.Owners.Count <= 1),
```
Comment (cristinamanum) [inline src/NuGetGallery/Controllers/UsersController.cs]: Why deleted?

File: src/NuGetGallery/Strings.resx [hunk 4]
```diff
@@ -649,4 +649,10 @@ For more information, please contact '{2}'.</value>
   <data name="UserPackageDeleteSupportRequestMessage" xml:space="preserve">
     <value>This support request is intended for automatic package deletion.</value>
   </data>
+  <data name="AccountDelete_OrganizationDeleteNotImplemented" xml:space="preserve">
+    <value>Account '{0}' cannot be deleted because it is an organization. A manual delete is required at this time.</value>
```
Comment (cristinamanum) [inline src/NuGetGallery/Strings.resx]: I think that the message needs to be something on the line:

 Account '{0}' cannot be deleted because it is an organization. The user needs to be manually removed from his organizations before the account to be deleted.

File: src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs [hunk 5]
```diff
@@ -2,19 +2,32 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System.Collections.Generic;
+using System.Linq;
 
 namespace NuGetGallery
 {
     public class DeleteAccountViewModel
     {
+        private bool? _hasOrphanPackages;
```
Comment (cristinamanum) [inline src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs]: Why need to be nullable?

Comment (cristinamanum) [reply src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs]: Oh, I see below.

File: src/NuGetGallery/Views/Users/_UserPackagesListForDeletedAccount.cshtml [hunk 6]
```diff
@@ -21,13 +21,13 @@
...
+                                @if (package.HasSingleOwner)
                                 {
-                                    <i  class="ms-Icon ms-Icon--CompletedSolid green-check" title="After account delete the package will still at least one owner."></i>
+                                    <i class="ms-Icon ms-Icon--StatusErrorFull red-check" title="After account delete the package will not have any owners."></i>
                                 }
-                                else 
+                                else
                                 {
-                                    <i  class="ms-Icon ms-Icon--StatusErrorFull red-check" title="After account delete the package will not have any owners."></i>
+                                    <i class="ms-Icon ms-Icon--CompletedSolid green-check" title="After account delete the package will still at least one owner."></i>
```
Comment (cristinamanum) [inline src/NuGetGallery/Views/Users/_UserPackagesListForDeletedAccount.cshtml]: will still have

**Comment by chenriksson** [comment]
Adding screenshots now:

Route '/account/delete' showing org packages (org has 2 members / packages not orphans):
image

Route '/account/Packages' showing org packages:
image

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 3 above)
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: The original orphan logic (count < 1) was both here and in _UserPackagesListForDeletedAccount.cshtml.

For the per-package orphan logic, I created the ListPackageItemViewModel.HasSingleOwner property. Property DeleteAccountViewModel.HasOrphanPackages still exists, but I now populate it based on the logic in ListPackageItemViewModel.HasSingleOwner... so I don't have duplicate code.

File: src/NuGetGallery/Extensions/UserExtensions.cs (hunk 1 above)
Comment (chenriksson) [reply src/NuGetGallery/Extensions/UserExtensions.cs]: True... I just use it to identify owners for a user. I'll try to find a better name.

File: src/NuGetGallery/Services/DeleteAccountService.cs (hunk 2 above)
Comment (chenriksson) [reply src/NuGetGallery/Services/DeleteAccountService.cs]: I opened #5025 for the remaining work, and linked it to #5016 which I'll use to track this initial work.

File: src/NuGetGallery/Strings.resx (hunk 4 above)
Comment (chenriksson) [reply src/NuGetGallery/Strings.resx]: Updating messages - will be more specific about what needs to be manually done before the account can be deleted.

File: src/NuGetGallery/ViewModels/ManagePackagesViewModel.cs [hunk 7]
```diff
@@ -7,6 +7,8 @@ namespace NuGetGallery
 {
     public class ManagePackagesViewModel
     {
+        public IEnumerable<string> PackageOwners { get; set; }
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManagePackagesViewModel.cs]: Remove this, which was intended for ManagePackages filtering... the next PR.

File: src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs [hunk 8]
```diff
@@ -2,19 +2,32 @@
...
 
         public User User { get; set; }
 
         public string AccountName { get; set; }
 
-        public bool HasOrphanPackages { get; set; }
-
         public bool HasPendingRequests { get; set; }
+
+        public bool HasOrphanPackages
```
Comment (skofman1) [inline src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs]: is this covered by UTs?

Comment (chenriksson) [reply src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs]: Added

File: src/NuGetGallery/ViewModels/ListPackageItemViewModel.cs [hunk 9]
```diff
@@ -49,6 +49,18 @@ public bool UseVersion
...
         }
 
+        public bool HasSingleOwner
+        {
+            get
+            {
+                var organization = Owners.FirstOrDefault() as Organization;
+
+                return Owners.Count == 1 &&
+                    // last owner is not an organization, or is an organization with one member.
```
Comment (scottbommarito) [inline src/NuGetGallery/ViewModels/ListPackageItemViewModel.cs]: I would rephrase this as
```
// A package has a single owner if its last owner is a user account or is an organization account that has a single member.
```

File: src/NuGetGallery/Extensions/UserExtensions.cs [hunk 10]
```diff
@@ -70,6 +71,22 @@ public static bool IsOwnerOrMemberOfOrganizationOwner(this User user, PackageReg
             return package.Owners.Any(o => user.KeyIsSelfOrOrganization(o.Key));
         }
 
+        /// <summary>
+        /// Get account keys for the user and their organizations.
+        /// This method is compatible with EF queries and can be used to filter packages by owner.
+        /// </summary>
+        public static List<int> GetUserAndOrganizationAccountKeys(this User user)
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Extensions/UserExtensions.cs]: Updated

File: src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs [hunk 11]
```diff
@@ -2,19 +2,32 @@
...
+        {
+            get
+            {
+                if (!_hasOrphanPackages.HasValue)
+                {
+                    _hasOrphanPackages = Packages?.Any(p => p.HasSingleOwner);
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/DeleteAccountViewModel.cs]: Updated

File: src/NuGetGallery/ViewModels/ListPackageItemViewModel.cs (hunk 9 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ListPackageItemViewModel.cs]: Updated

File: src/NuGetGallery/Services/PackageService.cs [hunk 12]
```diff
@@ -224,12 +227,16 @@ public IEnumerable<Package> FindPackagesByOwner(User user, bool includeUnlisted)
             return mergedResults.Values;
         }
 
+        /// <summary>
+        /// Merge latest stable packages by owner, including organization owners that the user belongs to.
+        /// </summary>
         private void MergeLatestStablePackagesByOwner(User user, bool includeUnlisted, Dictionary<string, Package> mergedResults)
         {
+            var ownerKeys = user.GetUserAndOrganizationAccountKeys();
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Services/PackageService.cs]: Updated

File: src/NuGetGallery/Services/PackageService.cs [hunk 13]
```diff
@@ -211,6 +211,9 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
             return package;
         }
 
+        /// <summary>
+        /// Find packages by owner, including organization owners that the user belongs to.
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Services/PackageService.cs]: Updated FindPackagesByOwner tests to include organization ownerships.

Regarding the debate about whether to update this method or the callers... since all callers (DeleteAccount, ManagePackages) need to include organization packages, I think it makes sense to update the API. The same will be true for FindPackageRegistrationsbyOwner, used by reserved namespace logic (#5017).

I'm not sure the rename is necessary, as it's still a lookup by owner. The only thing that changed is that a user can now have direct or indirect ownership. If we ever have callers that only want to look at direct ownership, we can update the APIs at that time.

[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Services/PackageService.cs]: Merging, but feel free to discuss P2P if you disagree.

[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Services/PackageService.cs]: @scottbommarito @shishirx34 Per discussion, sounds like the FindPackageRegistrationsByOwner should be fine as-is. I'll repurpose #5017 for developing a test plan so we can have a bug bash.
